File: app/src/main/java/com/android/lowes/requests/WeatherApiClient.java
package com.android.lowes.requests;

import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.android.lowes.AppExecutors;
import com.android.lowes.models.WeatherDetails;
import com.android.lowes.models.WeatherResponse;
import com.android.lowes.requests.responses.WeatherSearchResponse;
import com.android.lowes.util.Constants;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;

import retrofit2.Call;
import retrofit2.Response;

import static com.android.lowes.util.Constants.NETWORK_TIMEOUT;

public class WeatherApiClient {
    private static final String TAG = "WeatherApiClient";
    private static WeatherApiClient instance;


//    private MutableLiveData<List<WeatherResponse>> mWeathers;
private MutableLiveData<List<WeatherDetails>> mWeathers;
    private RetrieveWeatherRunnable mRetreiveWeatherRunnable;


    public static WeatherApiClient getInstance(){
        if(instance == null){
            instance = new WeatherApiClient();
        }
        return instance;
    }

    private WeatherApiClient() {
        mWeathers = new MutableLiveData<>();
    }

//    public LiveData<List<WeatherResponse>> getWeathers(){
//        return mWeathers;
//    }
    public LiveData<List<WeatherDetails>> getWeathers(){
        return mWeathers;
    }

    public void searchRecipesApi(String query){
        if(mRetreiveWeatherRunnable != null){
            mRetreiveWeatherRunnable = null;
        }
        mRetreiveWeatherRunnable = new RetrieveWeatherRunnable(query);
        final Future handler = AppExecutors.get().networkIO().submit(mRetreiveWeatherRunnable);

        // Set a timeout for the data refresh
        AppExecutors.get().networkIO().schedule(new Runnable() {
            @Override
            public void run() {
                // let the user know it timed out
                handler.cancel(true);
            }
        }, NETWORK_TIMEOUT, TimeUnit.MILLISECONDS);
    }

    private class RetrieveWeatherRunnable implements Runnable{

        private String query;

        private boolean cancelRequest;

        private RetrieveWeatherRunnable(String query) {
            this.query = query;
            cancelRequest = false;
        }

        @Override
        public void run() {

            try {
                Response response = getWeathers(query).execute();
                if(cancelRequest){
                    return;
                }
                if(response.code() == 200){
                    List<WeatherDetails> list = new ArrayList<>(((WeatherSearchResponse)response.body()).getWeathers());

                      //  List<WeatherDetails> currentWeather = mWeathers.getValue();

                     //   currentWeather.addAll(list);
                        mWeathers.postValue(list);

                }
                else{
                    String error = response.errorBody().string();
                    Log.e(TAG, "run: error: " + error);
                    mWeathers.postValue(null);
                }
            } catch (Exception e) {
                e.printStackTrace();
                mWeathers.postValue(null);
            }
        }

        private Call<WeatherSearchResponse> getWeathers(String query){
            return ServiceGenerator.getWeatherApi().searchWeather(
                    query,
                    Constants.API_KEY
            );
        }

        private void cancelRequest(){
            Log.d(TAG, "cancelRequest: canceling the retrieval query");
            cancelRequest = true;
        }
    }
}

File: app/src/main/java/com/android/lowes/requests/WeatherApiClient.kt
package com.android.lowes.requests

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.android.lowes.AppExecutors
import com.android.lowes.models.WeatherDetails
import com.android.lowes.requests.responses.WeatherSearchResponse
import com.android.lowes.util.Constants
import retrofit2.Call
import retrofit2.Response
import java.util.*
import java.util.concurrent.TimeUnit

class WeatherApiClient private constructor() {
    private  val TAG = "WeatherApiClient"
    private val mWeathers: MutableLiveData<List<WeatherDetails>?>
    private var mRetreiveWeatherRunnable: RetrieveWeatherRunnable? = null
    val weathers: LiveData<List<WeatherDetails>?>
        get() = mWeathers

    fun searchRecipesApi(query: String) {
        if (mRetreiveWeatherRunnable != null) {
            mRetreiveWeatherRunnable = null
        }
        mRetreiveWeatherRunnable = RetrieveWeatherRunnable(query)
        val handler = AppExecutors.get()?.networkIO()?.submit(mRetreiveWeatherRunnable)

        // Set a timeout for the data refresh
        AppExecutors.get()?.networkIO()?.schedule({ // let the user know it timed out
            handler?.cancel(true)
        }, Constants.NETWORK_TIMEOUT.toLong(), TimeUnit.MILLISECONDS)
    }

    private inner class RetrieveWeatherRunnable public constructor(private val query: String) : Runnable {
        private var cancelRequest = false
        override fun run() {
            try {
                val response: Response<*> = getWeathers(query)!!.execute()
                if (cancelRequest) {
                    return
                }
                if (response.code() == 200) {
                    val list: List<WeatherDetails> = ArrayList((response.body() as WeatherSearchResponse?)!!.weathers!!)
                    mWeathers.postValue(list)
                } else {
                    val error = response.errorBody()!!.string()
                    Log.e(TAG, "run: error: $error")
                    mWeathers.postValue(null)
                }
            } catch (e: Exception) {
                e.printStackTrace()
                mWeathers.postValue(null)
            }
        }

        private fun getWeathers(query: String): Call<WeatherSearchResponse?>? {
            return ServiceGenerator.weatherApi.searchWeather(
                    query,
                    Constants.API_KEY
            )
        }

        private fun cancelRequest() {
            Log.d(TAG, "cancelRequest: canceling the retrieval query")
            cancelRequest = true
        }

    }


    init {
        mWeathers = MutableLiveData()
    }

    private object HOLDER {
        val INSTANCE = WeatherApiClient()
    }

    companion object {
        val instance: WeatherApiClient by lazy { HOLDER.INSTANCE }
    }
}